purchases/src/main/kotlin/com/revenuecat/purchases/utils/SerialRequestExecutor.kt

```diff
@@ -0,0 +1,67 @@
+package com.revenuecat.purchases.utils
+
+import java.util.ArrayDeque
+import java.util.concurrent.atomic.AtomicBoolean
+
+internal typealias SerialRequest = (finish: () -> Unit) -> Unit
+
+internal class SerialRequestExecutor {
+
+    private val lock = Any()
+    private val pendingRequests = ArrayDeque<SerialRequest>()
+    private var hasActiveRequest: Boolean = false
+
+    /**
+     * Queues [request] to run after any in-flight request finishes. The provided [finish] callback
+     * **must** be called exactly once by the request, from either its success or error path, to allow
+     * the next queued request to start.
+     */
+    fun executeSerially(request: SerialRequest) {
+        val shouldStartNext = synchronized(lock) {
+            pendingRequests.addLast(request)
+            if (!hasActiveRequest) {
+                hasActiveRequest = true
+                true
+            } else {
+                false
+            }
+        }
+
+        if (shouldStartNext) {
+            executeNextRequest()
+        }
+    }
+
+    private fun executeNextRequest() {
+        val nextRequest = synchronized(lock) {
+            pendingRequests.firstOrNull()
+        } ?: run {
+            synchronized(lock) { hasActiveRequest = false }
```

Comment:
There is this super small chance of a race condition of this getting set to false if we check in thread 1 that there is nothing in the `pendingRequests` and right after but before this block is executed, on thread 2 the synchronized block in the `executeSerially` method above is executed.

Could we unify this to the `synchronized` block above that gets the first pending request?

Reply:
If it can happen, it probably will at some point! Good catch, unified the synchronized block here: d4c961afd954cff218175fd8aeea3fd6283ee13c